### ISSUE TITLE
Charges out by One on first rub on sweater

### ISSUE THREAD

**Issue by terracoda** [phetsims/balloons-and-static-electricity#207]
I think there is a related issue to this issue. Need to re-investigate.

When I pick up one negative charge (the first one), I received no update. The first charge alert was skipped. Next alert was a full update.

I can do more testing, but I just wanted to note it down.

**Comment by jessegreenberg** [comment]
I just observed this in the testing context.  It doesn't happen every time, but here is a capture:
![capture](https://cloud.githubusercontent.com/assets/6396244/20493023/6b841e26-afd4-11e6-9760-f3a3c7e7f4df.PNG)

**Comment by jessegreenberg** [comment]
It seems to happen bot on the first rub and on subsequent rubs if the balloon moves too slowly onto the sweater to pick up charges.

**Comment by terracoda** [comment]
Awesome capture, @jessegreenberg! I was thinking of using the visual PDOM for testing for issue #208, too.

**Comment by jessegreenberg** [comment]
Thanks @terracoda!  I believe the issue here is that interaction alerts are generated when the user completes an interaction rather than on model change.  I think this is what is causing #208 as well.  The balloon moves for a few ms after the key is up and might pick up another charge after the description has already been set as the text for the aria-live element.

In order to fix this, the commit above sets the content on the description when the property changes (rather than the user interaction).  But this sounds much worse because the AT is trying to speak while the user is interacting with the sim.  This causes the AT to stutter and even prevents alerts from being spoken at all.

I placed the changes on a branch because of these drawbacks.

**Comment by terracoda** [comment]
@jessegreenberg, this is an interesting finding. Thanks for testing it out a potential solution. I think your intuition is likely correct.

**Comment by terracoda** [comment]
@jessegreenberg, in the case of the 1 charge issue, the screen reader never said, "On sweater. Yellow balloon picks up negative charges from sweater," which ideally is the description of the first rub.

However, I just tested again (with ChromeVox - only thing that works for me on my system, sorry). On the first rub, when I picked up 1 negative charge, or even a few negative charges, I only got "On sweater," and no mention of the change in charge, then the next rub, I got the full update of the balloon and the sweater. 

The amount of charge did change, though, so I am wondering why the dynamic description wasn't, "On sweater. Picks up negative charges from sweater." However, when I did it again a little faster, the full correct alert was announced.

I'm not sure what is out of synch :-)

**Comment by terracoda** [comment]
![description-testing-synching-a-bit-off](https://cloud.githubusercontent.com/assets/1902116/20528678/bde4dc7c-b0a7-11e6-8f92-7f6173a58e1e.png)

In the above screen shot. The description in the yellow boxes are in synch with the sim, but the ones in red are not. @jessegreenberg, would your theory explain this?

**Comment by terracoda** [comment]
@jessegreenberg, and I must say, this visual PDOM is a very useful design tool!!!

**Comment by terracoda** [comment]
@jessegreenberg, oops forgot one annotation in above comment. I added the speech bubble with the missing content.
![description-testing-synching-a-bit-off](https://cloud.githubusercontent.com/assets/1902116/20528846/5571b40c-b0a8-11e6-8f8d-4839f68613e7.png)

**Comment by jessegreenberg** [comment]
Excellent, thanks for outlining the problem lines in the PDOM @terracoda! The problem is in the implementation of the interaction alerts.  The lines that are boxed in red are updated at the end of an interaction (user has lifted up a key on the keyboard).  The lines that are outlined in yellow are linked directly to the model.  Linking the interaction alerts perfectly with the model creates the issues described in https://github.com/phetsims/balloons-and-static-electricity/issues/207#issuecomment-262119486.

**Comment by jessegreenberg** [comment]
The red boxes under 'Scene' summary and 'Yellow Balloon' should be easy to fix since these don't use aria-live.  I am not sure yet how to fix the missing description under the 'Interaction Alerts' section.

**Comment by terracoda** [comment]
@jessegreenberg, this is very interesting!
FYI, the missing string, "Picks up negative charges from sweater," is announced sometimes. I am not what causes it not to be announced along with "On sweater," in this case.

**Comment by jessegreenberg** [comment]
@terracoda I see, thanks.  That actually sense if https://github.com/phetsims/balloons-and-static-electricity/issues/207#issuecomment-262119486 is correct. If you are able to pick up a charge while WASD keys are still down, you should be able to hear the correct charge update.  If the balloon moves and picks up a charge for a couple of milliseconds after the WASD keys are released, it won't be included in the interaction alert.

**Comment by jessegreenberg** [comment]
After inspecting the events, I no longer think this is true.  The alert is clearly coming after the location changing, even when the charge description is off by one.

![errant](https://cloud.githubusercontent.com/assets/6396244/20818754/f12b54fc-b7ec-11e6-8eeb-81deefb33ae2.PNG)
![errant2](https://cloud.githubusercontent.com/assets/6396244/20818756/f2d8a3fe-b7ec-11e6-9911-0705b507d8ff.PNG)

**Comment by jessegreenberg** [comment]
Here is the order of events that is causing this to happen:

![capture](https://cloud.githubusercontent.com/assets/6396244/20842067/68d77f4a-b873-11e6-9e76-41a9d76c0bbd.PNG)

In the interaction, the balloon is dragged, then the description is anounced, then the balloon picks up a final charge.

**Comment by terracoda** [comment]
@jessegreenberg, so if I am following you correctly, the "charge picked up" event should come before the keyup event?

**Comment by jessegreenberg** [comment]
Right! So we just have to figure out why it is out of order.

**Comment by jessegreenberg** [comment]
Mostly fixed in the commit above.  The balloon was being dragged via the view step function while the charges where being picked up via the view step function, causing the descriptions to be out of sync with the visual.

Since the above commit I haven't been able to reproduce with the aria-live alerts.
![fixed1](https://cloud.githubusercontent.com/assets/6396244/20907872/1ffa03a4-bb0e-11e6-8929-f4bb0a59fe1a.PNG)

In the same situation, the description is still out of sync so a similar solution will have to be applied here:
![broken1](https://cloud.githubusercontent.com/assets/6396244/20907876/2a55ad12-bb0e-11e6-8b8c-cbedeb94464e.PNG)

**Comment by jessegreenberg** [comment]
Should be fixed in commit above - I am unable to reproduce after the above commits.

![capture](https://cloud.githubusercontent.com/assets/6396244/20908970/e4a6a76e-bb15-11e6-8a8c-ef417cda3f8b.PNG)

@terracoda if you hever happen to run into this issue again in later versions, please feel free to open a new issue.

**Comment by terracoda** [comment]
Awesome work @jessegreenberg!
